float32 A
float32 B
string lines